Update instructions (#4)
dm-vdo/bunsen: Getting_Started.md

# Getting Started with Bunsen

> This guide is intended to help you get your first Bunsen environment up and running.

Reviewer: ```suggestion
This guide is intended to help get your Bunsen environment up and running.
```

Author: This will be in my next rev.

Author: This is addressed in my recent pushes.

# Expectations

> This project is simply an ansible playbook that has a certain set of requirements that need to be satisfied moving in.

Reviewer: ```suggestion
This project is an ansible playbook with specific requirements that need to be satisfied before beginning.
```

Reviewer 2: not sure I like beginning. Beginning what?

Author: I will provide my response in both threads for consideration.

```suggestion
This project is an ansible playbook with specific requirements that need to be satisfied before it can be used to provision an environment.
```

Reviewer: Good by me.

Reviewer 3: Good by me, as well.

> This project is simply an ansible playbook that has a certain set of requirements that need to be satisfied moving in.

Reviewer 3: I'm getting stuck on 'moving in'. Maybe change to 'before proceeding'?

Reviewer 2: before moving on i think is what he wants there

Author: The suggestion below from @jshimkus-rh is closer to what I had in my mind when writing this.

I will provide my response in both threads for consideration.

```suggestion
This project is an ansible playbook with specific requirements that need to be satisfied before it can be used to provision an environment.
```

Author: This is addressed in my recent pushes.

## Initial requirements

> Before you can start provisioning your environment, the following items must be satisfied:

Reviewer: ```suggestion
Before you can start provisioning your environment, the following criteria must be satisfied:
```

Author: This will be in my next rev.

Author: This is addressed in my recent pushes.

> 1. An environment must consist of at least 3 different systems (physical or virtual).  It is recommended that you have at least 5 machines, but minimally 3 machines can work.  More information on what the different systems will be doing will be provided later.

Reviewer: ```suggestion
1. At least 3 individual systems, physical and/or virtual.
    - It is recommended to have at least 5 systems.
    - The function of the separate systems are described below.
```

Reviewer 2: I don't know about this one. Isn't is sort of important to have set up the "systems" beforehand with some knowledge of what they will be used for? I'm thinking of needing disk space for the farms. Seems like just saying 3 or 5 here doesn't really give them much to go on. Maybe we need to be more specific ?

Author: Thank you.  I was having trouble with this one.

This will be in my next rev.

Author: This is addressed in my recent pushes.

>     - Of those 3 machines, one must have at least 280GiB of additional storage in a separate block device (i.e. sdb)

Reviewer: ```suggestion
    - At least one of the 3 systems must have a minimum of 280GiB of additional storage in a separate block device.
      - Generally this is /dev/sdb.
      - If the system is virtualized using 'kvm' it must be /dev/vdb.
```

Reviewer 2: This is the info I was talking about above. Maybe we don't need two places to discuss system count. Just here is enough?

Author: This will be in my next rev.

Author: This is addressed in my recent pushes.

> 2. Each machine must be able to communicate with the others in the environment both by IP address and by name (i.e. ossbunsen-resource [10.2.0.5] needs to be able to ssh to ossbunsen-farm-f36-1 [10.2.0.6])

Reviewer: ```suggestion
2. Each system must be able to communicate with the others in the environment both by IP address and by name (e.g., ossbunsen-resource [10.2.0.5] needs to be able to ssh to ossbunsen-farm-f36-1 [10.2.0.6])
```

Author: This will be in my next rev.

Author: This is addressed in my recent pushes.

3. The provisioning user must have an ssh-key generated and located under `~/.ssh`.
4. A `controller` system that has ansible installed on it.

You can find an example environment that utilizes vagrant and libvirt at [the VDO vagrant example repository](https://github.com/dm-vdo/vagrant)

## System Functions

> A provisioned Bunsen Environment consists of three different types of machine.

Reviewer: ```suggestion
A provisioned Bunsen Environment consists of three different classes of system.
```

Author: This will be in my next rev.

Author: This is addressed in my recent pushes.

- Infrastructure

>     - This machine will provide necessary infrastructure services for the other systems to utilize.

Reviewer: ```suggestion
    - This system provides necessary services for the other systems to utilize.
```

Author: This will be in my next rev.

Author: This is addressed in my recent pushes.

        - Such as NFS, [RSVP](https://github.com/dm-vdo/permabit-rsvpd) (resource management and coordination), iSCSI backing storage (if applcable)
- Resource

>     - General development and compiling will be done on this machine.

Reviewer: ```suggestion
    - General development and compiling will be done on these systems.
```

Author: This will be in my next rev.

Author: This is addressed in my recent pushes.

>     - If there are multiple incompatible targets (i.e. Fedora, RHEL), then one of each possible target should be available to ensure that the compilations are done according to their environment.

Reviewer: ```suggestion
    - If there are multiple incompatible targets (e.g., Fedora and RHEL), then one of each possible target should be available to ensure that the compilations are done according to their environment.
```

Author: This will be in my next rev.

Author: This is addressed in my recent pushes.

- Farm

>     - At least one of each applicable targets (i.e. Fedora, RHEL).

Reviewer: ```suggestion
    - At least one for each applicable target (e.g., Fedora or RHEL).
```

Author: This will be in my next rev.

Author: This is addressed in my recent pushes.

        - With Fedora, we typically utilize at least one of each supported version of Fedora in case there is a change applied to one and not the other.

>     - Each farm must have an additional disk (i.e. sdb) with 280 GiB or more of storage to use as scratch space for testing.

Reviewer: ```suggestion
    - Each farm must have an additional disk (see `Initial Requirements` for size and naming requirements) to use as scratch space for testing.
```

Author: This will be in my next rev.

Author: This is addressed in my recent pushes.

- Controller
    - This system can be the hypervisor that is running the various VMs.  Or it could be an additional VM living next to the others.

# Deploying your first environment

> Assuming that the dependencies above have been satisfied, you can now provision a Bunsen environment.  In order to deploy a bunsen environment, you need a few things first.

Reviewer: ```suggestion
Assuming that the dependencies above have been satisfied, you can now provision a Bunsen environment.  In order to deploy a bunsen environment, you need the following...
```

Author: A form of this will be in my next rev.  (probably without the trailing ...'s)

Author: This is addressed in my recent pushes.

> 1. First, we need a copy of this repository placed on a `controller` somewhere.

Reviewer: ```suggestion
1. A copy of this repository local to a `controller` system
```

Author: This will be in my next rev.

    - This system is not part of the minimal environment specified above.

> 2. Next, we need to generate an inventory file for ansible to utilize when provisioning the hosts.

Reviewer: ```suggestion
2. An inventory file for ansible to utilize when provisioning the hosts
```

Author: This will be in my next rev.

Author: This is addressed in my recent pushes.

    An example inventory file
    ```
    ossbunsen-infra      ansible_user=vagrant ansible_python_interpreter=/usr/bin/python3 ansible_ssh_pass=vagrant
    ossbunsen-resource   ansible_user=vagrant ansible_python_interpreter=/usr/bin/python3 ansible_ssh_pass=vagrant
    ossbunsen-farm-f36-1 ansible_user=vagrant ansible_python_interpreter=/usr/bin/python3 ansible_ssh_pass=vagrant
    
    [infrastructure]
    ossbunsen-infra
    
    [resources]
    ossbunsen-resource
    
    [farms]
    ossbunsen-farm-f36-1
    
    [all:vars]
    deploying_account = admin
    target_account = admin
    ```

    The `admin` account should be replaced with the account that you intend to do your development with.

>     NOTE: In place of `ansible_ssh_pass=vagrant`, you can use `ansible_ssh_private_key_file='/path/to/vagrantdir/.vagrant/machines/<name>/libvirt/private_key'` (of course if you're not using vagrant or you have another private key granted, then you should point at that instead)

Reviewer: ```suggestion
    NOTE: If using vagrant you can replace the use of `ansible_ssh_pass=vagrant` with `ansible_ssh_private_key_file='/path/to/vagrantdir/.vagrant/machines/<name>/libvirt/private_key'`.  If not using vagrant, or wanting to use a different private key, `ansible_ssh_private_key` should point to the appropriate private key.
```

Author: This will be in my next rev.

Author: This is addressed in my recent pushes.

    NOTE: You can also leave out `ansible_ssh_pass` and `ansible_ssh_private_key_file` and insted use `--ask-pass` when running the `ansible-playbook` commands below to be prompted for the passwords instead.

3. Now we can run the provisioning for the first time.  It is generally recommended to provision the roles serially to ensure that any dependencies across roles are satisfied without the chance of a timing issue being introduced.

    Example commands for each role.
    ```
    ansible-playbook -i inventory \
                     -l infrastructure \
                     provisioning/playbook.yml \
                     --extra-vars="install_bunsen_support=true" \
                     --ask-become-pass

    ansible-playbook -i inventory \
                     -l resources \
                     provisioning/playbook.yml

    ansible-playbook -i inventory \
                     -l farms \
                     provisioning/playbook.yml
    ```

    - Note the added options `--extra-vars="install_bunsen_support=true" --ask-become-pass` on the infrastructure role.

>         - `--extra-vars="install_bunsen_support=true"` instructs the ansible playbook to install the necessary COPR repositories on the `controller` system and install the python dependencies (python3-resource-discovery, and python3-utility-mill).  It is only needed to be specified one time.

Reviewer 3: Line 93: I'd suggest replacing the last sentence with "It only needs to be specified one time".

Author: This will be in my next rev.

Author: This is addressed in my recent pushes.

>         - `--extra-vars="install_bunsen_support=true"` instructs the ansible playbook to install the necessary COPR repositories on the `controller` system and install the python dependencies (python3-resource-discovery, and python3-utility-mill).  It is only needed to be specified one time.

Reviewer: ```suggestion
        - `--extra-vars="install_bunsen_support=true"` instructs the ansible playbook to install the necessary COPR repositories on the `controller` system and install the python dependencies (python3-resource-discovery, and python3-utility-mill).  It is required the first time you run the Bunsen playbook.  After that you may exclude using it unless you want to be certain to be using the latest version available; i.e., with it specified the most up-to-date version of the python dependencies will be installed.
```

Author: This will be in my next rev.

Author: This is addressed in my recent pushes.

>         - `--ask-become-pass` instructs the `ansible-playbook` command to prompr for the sudoers password on the `controller` to allow for you to install the packages.

Reviewer 3: Line 94: 'prompr' should be 'prompt'

Author: This will be in my next rev.

Author: This is addressed in my recent pushes.

        - If you would prefer NOT letting ansible provision your `controller`, please see the [Manually Installing Bunsen Support Libs](#installing-bunsen-support-libraries)
      
# Using the Bunsen environment

While the environment is yours and you can make it work as you wish, the most effective workflow involves doing general development on the `resource` machine and then doing the compilations and testing from that machine.

The home directory for the `target_account` specified above is mounted via NFS, so if you are doing work that requires multiple distributions, you don't have to unnecessarily duplicate things.

> To perform general development and testing for VDO, please see the VDO Getting Started document (This does not exist yet, contents TBD)

Reviewer 3: Line 103: Is this referring to this Getting_started.md, or something else? It's a bit confusing.

Author: Something else that does not exist yet.  That's what I was referring to in my comment above as hoping to write this week.

Author: This will be addressed at a later time (maybe tonight?) :)

# Getting Started with Bunsen

This guide is intended to help get your Bunsen environment up and running.

# Expectations

This project is an ansible playbook with specific requirements that need to be satisfied before it can be used to provision an environment.

## Initial requirements

Before you can start provisioning your environment, the following criteria must be satisfied:

1. At least 3 individual systems, physical and/or virtual.
    - It is recommended to have at least 5 systems.
    - The function of the separate systems are described below.
    - At least one of the 3 systems must have a minimum of 280 GiB of additional storage in a separate block device.
        - Generally this is /dev/sdb.

>         - If the system is virtualized using 'kvm', it can be /dev/vdb.

Reviewer: Bunsen file `provisioning/roles/farm/tasks/storage.yml` has the following:
```
      # Use the auxiliary disk, if it exists.
      - name: Setting scratch volume group and test disk names
        set_fact:
          scratch_vg: scratch
          test_disk: "/dev/{{aux_disk}}"

      vars:
        aux_disk: "{%- if ansible_virtualization_type == 'kvm' -%}
                     vdb
                   {%- else -%}
                     sdb
                   {%- endif -%}"
      when: aux_disk in ansible_devices.keys()
```

which is why my previous comment used "must" be /dev/vdb.  Based on the scenarios for setting `test_disk` (unless I missed something) it appears that if it's kvm and not vdb the provisioning will fail in
```
  - name: Verifying storage variables have been set
    assert:
      that:
        - scratch_vg is defined
        - test_disk is defined
```

Our general experience has been that the various virtualization technologies used automatically, in a two disk system, use either sdb or vdb.  Bunsen was written with that in mind.  Maybe it'd be better to make the description be that it must be a two drive system with the second disk being 280GiB, etc. rather than specifically mentioning device names.

Author: When I was calling out a particular device (sdb), it was really just to convey that it was another device, not the driver that was providing it.  Similarly, if you're on kvm, you can still use a different driver for the storage and end up with a `sdb` instead of `vdb`, but it's typical to get `vdb` if you're not messing with things.

I think I'd prefer the current spelling where we say it's generally going to be `/dev/sdb`, and if it's kvm that it can be `vdb`.  Perhaps I can rephrase it to mention using the `virtio driver` rather than kvm.

What do you think?

Reviewer: Simply having a second storage device is not sufficient for it to be used; for example, I can create a second storage device as NVMe and it's not touched.  The current implementation of bunsen has a number of fallbacks but they're generally predicated on there being enough free space on the root disk; if the root disk doesn't have enough free space provisioning will fail.

There are, as I see it, three choices:
1. extend bunsen to handle more scenarios w/o requiring specific device names
2. proscribe that in environment X the second device must be named Y
3. rely on our experiences that "the right thing" happens and do not mention any device name

The robustness of option 1 will probably always have some holes.
Option 2 is akin to any other environmental qualification; it might not meet some folks desires but they could then extend bunsen.
Option 3, if someone encountered a problem, could be explained away as we've been using default behavior of various virtualization tech; if they are specifying a non-default behavior that's their issue.  Again, they could extend bunsen.

The engineer in me would like to do option 1, but pragmatically I think either option 2 or 3 is the way to go at present.

Author: I think we need to hit the pause button on this discussion.

The primary purpose of this PR is to establish a set of instructions to  use; something we don't have posted yet.

So with all that being said, I think we should go with the easiest option and not be too strict about our spelling.  We can always adjust it after the fact.  If we are going to decide only through the three options you provided, then given these constraints, I would opt for option 3 and just omit the names and let people figure it out the hard way until someone with more time can address it more completely.

2. Each system must be able to communicate with the others in the environment both by IP address and by name (e.g. ossbunsen-resource [10.2.0.5] needs to be able to ssh to ossbunsen-farm-f36-1 [10.2.0.6])
3. The provisioning user must have an ssh-key generated and located under `~/.ssh`.
4. A `controller` system that has ansible installed on it.

You can find an example environment that utilizes vagrant and libvirt at [the VDO vagrant example repository](https://github.com/dm-vdo/vagrant)

## System Functions

A provisioned Bunsen Environment consists of three different classes of system.

- Infrastructure
    - This system provides necessary services for the other systems to utilize.

>         - Such as NFS, [RSVP](https://github.com/dm-vdo/permabit-rsvpd) (resource management and coordination), iSCSI backing storage (if applcable)

Reviewer: ```suggestion
        - Such as NFS, [RSVP](https://github.com/dm-vdo/permabit-rsvpd) (resource management and coordination), iSCSI backing storage (if applicable)
```

Author: This is fixed.
